### PR TITLE
udp_proxy: support proxy access logging

### PR DESCRIPTION
**Pull request by giantcroc**
Signed-off-by: giantcroc [changran.wang@intel.com](mailto:changran.wang@intel.com)

Commit Message: 
Some stats like `no_route` and `idle_timeout` can't be printed by session access log,
so we need proxy-level access logging to log global stats.
Additional Description:
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #23241
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto [hunk 1]
```diff
@@ -105,6 +105,9 @@ message UdpProxyConfig {
   // to upstream host selected on first chunk receival for that "session" (identified by source IP/port and local IP/port).
   bool use_per_packet_load_balancing = 7;
 
-  // Configuration for access logs emitted by the UDP proxy. Note that certain UDP specific data is emitted as :ref:`Dynamic Metadata <config_access_log_format_dynamic_metadata>`.
-  repeated config.accesslog.v3.AccessLog access_log = 8;
+  // Configuration for session access logs emitted by the UDP proxy. Note that certain UDP specific data is emitted as :ref:`Dynamic Metadata <config_access_log_format_dynamic_metadata>`.
+  repeated config.accesslog.v3.AccessLog sess_access_log = 8;
```
Comment (lizan) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:109]: Prefer not abbreviate, use `session_access_log`

**Review by lizan — COMMENTED** [review]
API review

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto [hunk 2]
```diff
@@ -105,6 +105,9 @@ message UdpProxyConfig {
...
+
+  // Configuration for proxy access logs emitted by the UDP proxy. Note that certain UDP specific data is emitted as :ref:`Dynamic Metadata <config_access_log_format_dynamic_metadata>`.
+  repeated config.accesslog.v3.AccessLog proxy_access_log = 10;
```
Comment (daixiang0) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:112]: where is the "9"?

File: docs/root/configuration/observability/access_log/usage.rst [hunk 3]
```diff
@@ -650,28 +650,50 @@ The following command operators are supported:
 
   UDP
     For :ref:`UDP Proxy <config_udp_listener_filters_udp_proxy>`,
-    NAMESPACE should be always set to "udp.proxy", optional KEYs are as follows:
+    if NAMESPACE is set to "udp.proxy.session", optional KEYs are as follows:
```
Comment (daixiang0) [inline docs/root/configuration/observability/access_log/usage.rst:653]: `When NAMESPACE ...`

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto (hunk 2 above)
Comment (giantcroc) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:112]: It is already used at here `xds.type.matcher.v3.Matcher matcher = 9`

Comment (daixiang0) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:112]: Oh, I miss it, better to make them in order.

File: docs/root/configuration/observability/access_log/usage.rst (hunk 3 above)
Comment (daixiang0) [reply docs/root/configuration/observability/access_log/usage.rst:653]: Maybe `NAMESPACE should be always set to "udp.proxy.session"`

File: docs/root/configuration/observability/access_log/usage.rst [hunk 4]
```diff
@@ -650,28 +650,50 @@ The following command operators are supported:
...
+    if NAMESPACE is set to "udp.proxy.proxy", optional KEYs are as follows:
+
+    * ``bytes_sent``: Total number of downstream bytes sent to the upstream in UDP proxy.
+    * ``bytes_received``: Total number of downstream bytes received from the upstream in UDP proxy.
+    * ``errors_sent``: Number of errors that have occurred when sending datagrams to the upstream in UDP proxy.
+    * ``errors_received``: Number of errors that have occurred when receiving datagrams from the upstream in UDP proxy.
+    * ``datagrams_sent``: Number of datagrams sent to the upstream successfully in UDP proxy.
+    * ``datagrams_received``: Number of datagrams received from the upstream successfully in UDP proxy.
+    * ``no_route``: Number of times that no upstream cluster found in UDP proxy.
+    * ``sess_total``: Total number of sessions in UDP proxy.
```
Comment (daixiang0) [inline docs/root/configuration/observability/access_log/usage.rst:682]: sesson_total?

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc [hunk 5]
```diff
@@ -281,29 +294,51 @@ UdpProxyFilter::ActiveSession::~ActiveSession() {
...
+  fields_map["errors_received"] =
+      ValueUtil::numberValue(config_->stats().downstream_sess_rx_errors_.value());
+  fields_map["datagrams_sent"] =
+      ValueUtil::numberValue(config_->stats().downstream_sess_tx_datagrams_.value());
+  fields_map["datagrams_received"] =
+      ValueUtil::numberValue(config_->stats().downstream_sess_rx_datagrams_.value());
+  fields_map["no_route"] =
+      ValueUtil::numberValue(config_->stats().downstream_sess_no_route_.value());
+  fields_map["sess_total"] =
+      ValueUtil::numberValue(config_->stats().downstream_sess_total_.value());
```
Comment (daixiang0) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:338]: A full name like `downstream_session_total` may be better? Is there a length limit?

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h [hunk 6]
```diff
@@ -116,7 +121,12 @@ class UdpProxyFilterConfig {
   const Network::ResolvedUdpSocketConfig& upstreamSocketConfig() const {
     return upstream_socket_config_;
   }
-  const std::vector<AccessLog::InstanceSharedPtr>& accessLogs() const { return access_logs_; }
+  const std::vector<AccessLog::InstanceSharedPtr>& sessionAccessLogs() const {
+    return sess_access_logs_;
```
Comment (daixiang0) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:125]: full name here?

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto (hunk 2 above)
Comment (giantcroc) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:112]: Ok, I'll do this by the way.

File: docs/root/configuration/observability/access_log/usage.rst (hunk 3 above)
Comment (giantcroc) [reply docs/root/configuration/observability/access_log/usage.rst:653]: > `When NAMESPACE ...`

OK, done, thanks!

Comment (giantcroc) [reply docs/root/configuration/observability/access_log/usage.rst:653]: > when `NAMESPACE is set to "udp.proxy.session"`

when users want to log stats recorded in sessions. And it makes sure the tokens like `bytes_sent` can be reused.

File: docs/root/configuration/observability/access_log/usage.rst (hunk 4 above)
Comment (giantcroc) [reply docs/root/configuration/observability/access_log/usage.rst:682]: OK, fixed, thanks!

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc (hunk 5 above)
Comment (giantcroc) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:338]: I think `session_total` is only meaningful to downstream, so there is no need to add `downstream` prefix for simplicity.

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h (hunk 6 above)
Comment (giantcroc) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:125]: OK, fixed, thanks!

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto (hunk 1 above)
Comment (giantcroc) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:109]: OK, done, thanks!

**Comment by giantcroc** [comment]
@phlax Hi, it seems wierd that the old code is still used to do test
```
test/extensions/filters/udp/udp_proxy/udp_proxy_filter_test.cc:639:42: error: too few arguments to function call, expected 3, have 2
                        access_log_format));
```
Can you help with it? Thanks!

**Comment by lizan** [comment]
@giantcroc CI runs on merge commit and you can see that part code here: https://github.com/envoyproxy/envoy/blob/e143e084f5/test/extensions/filters/udp/udp_proxy/udp_proxy_filter_test.cc#L629-L639

It's a merge race with #23121, you just need to merge main branch and fix that test as well.

**Comment by giantcroc** [comment]
> @giantcroc CI runs on merge commit and you can see that part code here: https://github.com/envoyproxy/envoy/blob/e143e084f5/test/extensions/filters/udp/udp_proxy/udp_proxy_filter_test.cc#L629-L639
> 
> It's a merge race with #23121, you just need to merge main branch and fix that test as well.

OK, fixed, thanks a lot!

**Comment by lizan** [comment]
The coverage failure is a real one so you either need increase coverage or change threshold.

**Comment by giantcroc** [comment]
> The coverage failure is a real one so you either need increase coverage or change threshold.

Hi, lizan!  Sorry for late respond. The coverage has been fixed after rebase. Thanks!

**Review by lizan — APPROVED** [review]

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto (hunk 2 above)
Comment (lizan) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:112]: Never ever renumber fields...

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto [hunk 7]
```diff
@@ -66,13 +66,13 @@ message UdpProxyConfig {
 
     // The match tree to use when resolving route actions for incoming requests.
     // See :ref:`Routing <config_udp_listener_filters_udp_proxy_routing>` for more information.
-    xds.type.matcher.v3.Matcher matcher = 9
+    xds.type.matcher.v3.Matcher matcher = 3
```
Comment (lizan) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:69]: please revert all the re-numbering and only add new field with `10`, otherwise it break backwards compatibilty.

**Review by lizan — CHANGES_REQUESTED** [review]

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto (hunk 7 above)
Comment (giantcroc) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto]: OK, fixed, thanks!

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto [hunk 8]
```diff
@@ -105,6 +105,9 @@ message UdpProxyConfig {
   // to upstream host selected on first chunk receival for that "session" (identified by source IP/port and local IP/port).
   bool use_per_packet_load_balancing = 7;
 
-  // Configuration for access logs emitted by the UDP proxy. Note that certain UDP specific data is emitted as :ref:`Dynamic Metadata <config_access_log_format_dynamic_metadata>`.
-  repeated config.accesslog.v3.AccessLog access_log = 8;
+  // Configuration for session access logs emitted by the UDP proxy. Note that certain UDP specific data is emitted as :ref:`Dynamic Metadata <config_access_log_format_dynamic_metadata>`.
+  repeated config.accesslog.v3.AccessLog session_access_log = 8;
```
Comment (lizan) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:109]: Don't rename existing field, this breaks backward compatibility as well.

Please read https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md#backwards-compatibility

Comment (giantcroc) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:109]: OK, but I think we should distinguish different fields that for session and proxy, so maybe I could add a behavior change log.

**Review by ZiyangXiao — COMMENTED** [review]
Should remove this line in this pr?
https://github.com/envoyproxy/envoy/blob/82667905a934826576e817a8222b0e1117b79b15/source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h#L28

**Comment by giantcroc** [comment]
> Should remove this line in this pr?
> 
> https://github.com/envoyproxy/envoy/blob/82667905a934826576e817a8222b0e1117b79b15/source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h#L28

OK, thanks!

**Comment by giantcroc** [comment]
@lizan Hi, I have recovered the field name to address compatibility issue. Please help review, thanks!

File: changelogs/current.yaml [hunk 9]
```diff
@@ -39,6 +39,56 @@ new_features:
 
 - area: http
   change: |
+    added the expected :ref:`receive <envoy_v3_api_field_config.core.v3.HealthCheck.HttpHealthCheck.receive>` payload check for HTTP health check.
```
Comment (lizan) [inline changelogs/current.yaml:42]: there's a lot of unrelated changes in this version history which might comes from resolving conflicts, please merge main again and add only what you did in this PR.

Comment (giantcroc) [reply changelogs/current.yaml]: OK, fixed, thanks!

**Review by lizan — APPROVED** [review]
